iris-mpc-cpu/src/execution/hawk_main/state_check.rs

```diff
@@ -0,0 +1,112 @@
+use eyre::{eyre, Result};
+use itertools::{chain, Itertools};
+use serde::{Deserialize, Serialize};
+use siphasher::sip::SipHasher13;
+use std::hash::{Hash, Hasher};
+
+use super::{HawkSession, HawkSessionRef};
+
+#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
+pub struct SetHash {
+    accumulator: u64,
+}
+
+impl SetHash {
+    pub fn add_unordered(&mut self, value: impl Hash) {
+        self.accumulator = self.accumulator.wrapping_add(Self::hash(value));
+    }
+
+    pub fn remove(&mut self, value: impl Hash) {
+        self.accumulator = self.accumulator.wrapping_sub(Self::hash(value));
+    }
+
+    pub fn digest(&self) -> u64 {
+        self.accumulator
+    }
+
+    fn hash(value: impl Hash) -> u64 {
+        let mut hasher = SipHasher13::default();
+        value.hash(&mut hasher);
+        hasher.finish()
+    }
+}
```

Comment:
This could be very easy to misuse IMO, since it just takes all impl Hash, which do not enforce the correct usage on call sites. It is fine RN since the interface is actually pretty small, but could it make sense to have it be something like
```rust
struct SetHash<T: Hash> {
  accumulator: u64,
  phantom: PhantomData<T>
 }
 impl<T: Hash> SetHash<T> {
   pub fn add_unordered(&mut self, value: T {
        self.accumulator = self.accumulator.wrapping_add(Self::hash(value));
    }
    ...
 }
 ```
 
 And restrict a single instance of the SetHash to be for one type only?
 
 Also can probably just take references here instead to avoid moving the hashed value in here.

Reply:
This can be more type safe, although I do use it for fields of different types (graph entry point and layers), and writing the T can get verbose (enums, lifetimes, etc).

It does work by reference or tuples of references.